src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/loptimizer/heuristics/formula/JoinAwareWeightedUnboundVariableCount.java

```diff
@@ -0,0 +1,113 @@
+ * "Heuristics-based Query Reordering for Federated Queries in SPARQL 1.1 and SPARQL-LD"
+ */
+
+public class JoinAwareWeightedUnboundVariableCount {
+    protected static final double W_S = 1; // weight for subject variables
+    protected static final double W_P = 0.1; // weight for predicate variables
+    protected static final double W_O = 0.8; // weight for object variables
+
+    protected static final double J_Tc = 0.6; // weight for chain join
+    protected static final double J_Ts = 0.5; // weight for star join
+    protected static final double J_Tu = 1; // weight for unusual join
+
+    public static double estimate( final List<Query_Analyzer> selectedPlans, final Query_Analyzer subPlan ) {
+        // Add bound variables of all selected plans to a set
+        Set<Node> boundVariables = new HashSet<>();
+
+        // This part can be optimized if all formulas need to consider bound variables:
+        // adding to this set directly when adding a new subPlan to 'selectedPlans'
+        for ( Query_Analyzer plan : selectedPlans ) {
+            boundVariables = addBinds( plan.getSubs(), plan.getPreds(), plan.getObjs() );
+        }
+
+        return calculateCost( boundVariables, subPlan );
+    }
+
+    // Formula (7) in paper "Heuristics-based Query Reordering for Federated Queries in SPARQL 1.1 and SPARQL-LD"
+    private static double calculateCost( final Set<Node> bindings, final Query_Analyzer subPlan ) {
+        final double unboundVarsCost = getUnboundVarsCost(
+                subPlan.getSubs(),
+                subPlan.getPreds(),
+                subPlan.getObjs(),
+                bindings);
+        final double joinCost = joinsWeight(
+                subPlan.getSubs(),
+                subPlan.getPreds(),
+                subPlan.getObjs());
+        return unboundVarsCost / joinCost;
+    }
+
+    private static double getUnboundVarsCost( final List<Node> vars_s, final List<Node> vars_p,
+                                              final List<Node> vars_o, final Set<Node> boundVariables) {
+        final Set<Node> varsTotal = new HashSet<>();
+        // Calculate the number of (unique) unbound subjects, predicates and objects
+        final int totalSubs = calculateVars(vars_s, varsTotal, boundVariables);
+        final int totalObjs = calculateVars(vars_o, varsTotal, boundVariables);
+        final int totalPreds = calculateVars(vars_p, varsTotal, boundVariables);
+
+        return calculateTripleWeights(totalSubs, totalPreds, totalObjs);
+    }
+
+    /**
+     * Creates a collection with all variables that have been currently calculated
+     * and adds the bound variables to bindings
+     * @param vars A list of variables, can be subs, preds or objects
+     * @param varsTotal A list of all variables that have been currently calculated in this subquery
+     * @param boundVariables All bound variables (including variables in selected plans and calculated part of this subquery)
```

Comment:
Also here, I don't understand what you mean by "calculated". Do you want to say "counted"?